README
1.  ./autogen.sh
    ./configure --host=i686-linux-gnu "CFLAGS=-m32" "CXXFLAGS=-m32" "LDFLAGS=-m32"
    make
    make install

2.  Make certain you have the file, rpgd.conf, in the directory /etc.  There
    is a sample rpgd.conf in the rpgd/etc that has in it:

HOME=/usr/local/rpgd
GROUP=nobody
USER=nobody
#DEV=/dev/ttyS1
#ROWS=4
#COLS=20

HOME, GROUP, and USER are required for rpgd to run.  rpgd must run as 'root',
but will change its current work directory to HOME, and its user and group id
to whatever is specified (nobody).

DEV, ROWS, and COLS are optional parameters to use to dump user logoff 
statistics to, i.e., to a serial video display in your server.  The # is
used to comment out this feature.

3.  Make certain rpgd is running.  It should install a line in your server's
    /etc/inittab, like the following:

# Run rpgd
rpgd:35:respawn:/usr/local/bin/rpgd -d

You might have to issue a 'telinit q' at the root console prompt to inform
init that its configuration file has been changed and needs to be re-read.

    For init hosts, copy init/rpgd.conf to /etc/init

    For systemd hosts, copy etc/rpgd.service to
    /etc/systemd/system/multi-user.target.wants

Check for rpgd by typing:

ps auxw | grep rpg

You can check its log:

cat log/rpgd.log

It should start with:
rpgd v2.5.7
(C)1999-2014 Robert Hurst
All Rights Reserved.


4.  Run the rpgclient program or use the 'hack' convenience script.
    If you get a 'connection refused' error message, that means rpgd is NOT
    running.

5.  Connecting remotely with the Windows RPGCLIENT.exe program.  By default,
    rpgclient is on port #7000 and is inserted in your /etc/services file:

grep rpg /etc/services

rpgclient       7000/tcp                        # RPGclient

If that line is not in there, you cannot connect remotely.

By default, an /etc/xinetd.d/rpgclient service script is created:

cat /etc/xinetd.d/rpgclient

# default: off
# description: The RPGD client allows remote users to login to play
#              Hack & Slash.
service rpgclient
{
        disable = no
        socket_type             = stream
        wait                    = no
        user                    = root
        server                  = /usr/local/bin/rpgclient
        log_on_success          += USERID
        log_on_failure          += USERID
}

You might have to restart the xinetd service, such as:

service xinetd restart

6.  Make certain your firewall allows port #7000 to come in.

7.  Make certain that the rpgd binaries are installed in a directory that is
    part of the root's path, i.e., /usr/local/bin

echo $PATH

/usr/kerberos/sbin:/usr/kerberos/bin:/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/X11R6/bin:/usr/local/j2sdk1.4.1/bin:/usr/local/Sophos/bin:/root/bin


Happy Hunting!



ChangeLog
      ---------
--=:)) HISTORY ((:=--
      ---------

----------
Linux RPGD
----------
04-Jul-2011   * Updated for init and systemd hosts.  Uses Anjuta IDE with
                autogen.  Some minor cosmetic improvements.

20-Sep-2008   * Major bug-fix on USER master file; it is now 64-bit friendly

23-Aug-2008   * Updated all source to allow compiling with gcc 4.3
                Removed makefiles and replaced with Code::Blocks project files

24-Apr-2005   * Added WILD icon and adjusted payoffs in Casino's Slot Machine.
                Corrected transmission of soundbytes.  Added cancel typeahead
                features in critical areas to avoid random events from taking
                buffered keyboard events, such as Y/N prompts and falling 
                through dungeon trapdoors.  Added some honor among thieves in
                the dungeon.

11-Feb-2004   * Added Big Blast to most memorable hits.  Adjusted Thief to
                an average melee with less backstab power.  Fixed skill from
                Wands to Scrolls to work.  Various bugs fixed.

23-Jan-2004   * Reaching level 50 now earns you a skill point to assign to
                your character, either +10 to any ability attribute or melee,
                backstab, poison, magic, or stealing.  Some class table mods
                made to starting points.

31-Dec-2004   * Tweaked character classes and skill sets.  New backstab
                factor in battle engine.

19-Apr-2003   * Fixed install script to create a default /etc/rpgd.conf

18-Apr-2003   * Minor console and sound bug fixes.  Added a 'hack' shell script
                that kills playing music and sets a cool console font before
                running the rpgclient.

21-Jul-2002   * Minor updates to cgi utility, rpgwatch, and package.

16-Aug-2001   * Fixed fork() _exit completion in sndplay() to comply with glibc
                spec.  Minor tweaks to characteristics and spoils to the deeper
                dungeon levels.

04-Mar-2001   * Updated documentation.

17-Feb-2001   * Ported all packages to the KDevelop C/C++ IDE.  Installation
                changes.

14-Jan-2001   * Rewrote AI within Battle() to be smarter in its spell casting
                selections, as well as tweaked excellent hits and blasts.
                Fixed treasure detect map.  Adjusted Wheel of Life Grace and
                Doom! payoffs.  Ogre-class has been redefined to be the best
                at melee, at further expense of dexterity.  Fixed potential
                midnight bug in daemon.

07-Jan-2001   * Changed New User registration to be more 'net friendly.  The
                inclusion of an e-mail address field now auto-validates the
                account.  This makes sendmail a prerequisite.  Extensive
                rewrite of rpgweb to include too many web features to list.
                Also added an enchanted map feature in the dungeon.  Updated
                GTK+ sysop.  Fixed cleanup routines.  Happy 11th Birthday to
                Erin J. Hurst.

01-Jan-2001   * Updated source & binaries to work correctly with Linux 2.2.18
                kernel and glibc-2.2.  Updated installation script to be
                xinetd aware.  Added rpgweb for web site output.

04-Sep-2000   * Fixed GTK Sysop build and some internal flags.  Added dungeon
                armor and weapon gift shops.  Enhanced the Legendary Wishing
                Well and Wheel of Life routines.

27-Aug-2000   * Added portals to other deep, dank dungeons.  Exploring them
                widens the range of risks & rewards, as well as a few more
                treats along the way.  Other minor bug fixes.

31-Jul-2000   * Minor bug fixes to rpgd server.  Changed bail-out from jail
                algorithm at login.  New player rating algorithms.  Added
                casino music.

05-May-2000   * Added many more sounds.

25-Mar-2000   * Fixed 5-card = 21 bug.  Rounded some money calculations to
                nearest coinage for a more realistic look.  Added a magic
                map treasure to each dungeon floor.

27-Feb-2000   * Added a bonus sound.  Fixed RPGclient packager.

29-Jan-2000   * Updated install script, a surprised dungeon thief will vanish,
                more sound events, better DEC OSF/1 support, and more minor
		bug fixes.

23-Jan-2000   * Improved async input handling with timeouts.  Updated GTK
                Sysop utility.  Fixed monster poisoning.

16-Jan-2000   * Created GTK Sysop utility.  Added toss vial option in Dungeon.
                Minor bug fixes.

14-Jan-2000   * Added better FreeBSD support [thanks Shane Bryldt for use of
	        his FreeBSD server].  Install/Makefiles updated.  term.c fixed.

11-Jan-2000   * Minor bug fixes that check Origin/Name of Class, Armor, and
                Weapon.  Bug fix when checking a new dungeon level.  Thanks to
                Travis Mikalson for his debugging.

06-Jan-2000   * Re-engineered install procedure.  Binaries are optimized to
	        less than half their original size.

30-Dec-1999   * Moved Won Game stuff from client to server, gives any online
                users at the time the boot.  Updated some Sysop functions.
                Got rid of visitors in the newspaper.  Keep last 5 callers,
                instead of 3.  Better signal trapping.

25-Dec-1999   * More robust client-server handling of timeouts.  Split out
                and display the effect of ego weapon/armor during battle.
                Added a check when adventuring in the dungeon to influence bad
                events upon the player without reasonable security for his
                experience level.  Wishing Well can generate multiple key
                hints and spells.  Fixed an overflow bug in rpgclient.exe.

19-Dec-1999   * Tax Collector monitors tavern entrance.  RPGD sends HUP signal
                to idle client after 10-minutes and KILL signal after 20-

include/rpgd.h
//	*******************************
//	*  RPGBBS Deluxe File Server  *
//	*                             *
//	*   Written by Robert Hurst   *
//	*******************************
#define	SERVER_SHUTDOWN		-1
#define	SERVER_CONNECT		 1
#define	SERVER_DISCONNECT	 2
#define	SERVER_BROADCAST	 3
#define	SERVER_DELUSER		 4
#define	SERVER_GETUSER		 5
#define	SERVER_NEXTUSER		 6
#define	SERVER_PREVUSER		 7
#define	SERVER_PUTUSER		 8
#define	SERVER_RNDUSER		 9
#define	SERVER_GETUSERLEVEL	10
#define	SERVER_GETUSERNAME	11
#define	SERVER_LOCK			12
#define	SERVER_UNLOCK		13
#define	SERVER_DELGANG		14
#define	SERVER_GETGANG		15
#define	SERVER_NEXTGANG		16
#define	SERVER_PREVGANG		17
#define	SERVER_PUTGANG		18
#define	SERVER_GETHISCORE	19
#define	SERVER_WONGAME		20

#define	STRENGTH_SPELL		1L<<0
#define	INTELLECT_SPELL		1L<<1
#define	DEXTERITY_SPELL		1L<<2
#define	CHARISMA_SPELL		1L<<3
#define	SHIELD_SPELL		1L<<4
#define	HONE_SPELL			1L<<5
#define	HEAL_SPELL			1L<<6
#define	TELEPORT_SPELL		1L<<7
#define	BLAST_SPELL			1L<<8
#define	RESURRECT_SPELL		1L<<9
#define	CONFUSION_SPELL		1L<<10
#define	TRANSMUTE_SPELL		1L<<11
#define	CURE_SPELL			1L<<12
#define	ILLUSION_SPELL		1L<<13
#define	DISINTEGRATE_SPELL	1L<<14
#define	MORPH_SPELL			1L<<15

#define	ARMOR_RUSTING_XSPELL	1<<0
#define	WEAPON_DECAY_XSPELL		1<<1
#define	BIG_BLAST_XSPELL		1<<2
#define	MANA_STEALING_XSPELL	1<<3
#define	LIFE_STEALING_XSPELL	1<<4
#define LEVEL_STEALING_XSPELL	1<<5
#define SUPER_SHIELD_XSPELL	1<<6
#define SUPER_HONE_XSPELL	1<<7

#define	USER_RECORD_SIZE		512L

struct charstats {
	UWORD Calls,Kills,Killed,Retreats,KOs,KOed,JoustWin,JoustLoss;
};

struct playerstats {
	ULONG HP,Hull,Kills;
	UWORD Calls,Killed,Retreats,KO,Joust,Stole,Gamble,Fish;
};

struct notable {
	BYTE Who[23];
	UWORD What;
	JDATE When;
};

struct class_character {
	BYTE Name[10];
	UBYTE Melee;		//  fighting ability: 0x-8x
	UBYTE Backstab;	//  0=1%, 1=5%, 2x-8x
	UBYTE Poison;		//  0=none, 1x-8x
	UBYTE Magic;		//  0=none, 1=wand, 2=scroll, 3=spell, 4=enchanted
	UBYTE Steal;		//  1x-8x
	UBYTE BaseSTR,BaseINT,BaseDEX,BaseCHA;
	UBYTE BonusSTR,BonusINT,BonusDEX,BonusCHA;
	UBYTE MaxSTR,MaxINT,MaxDEX,MaxCHA;
	struct notable Calls[2],Kills[2],Killed[2],Retreats[2],Brawls[2],Jousts[2],melee,blast,bigblast;
};
struct class {
	BYTE Origin[9];
	UBYTE NumCharacter;
	struct class_character **Character;
};

struct user {
	BYTE ID[5];
	BYTE Handle[23];
	BYTE Name[31];
	BYTE Email[51];
	UBYTE Filler[38];
	BYTE Password[10];
	JDATE BirthDate,CreateDate,ExpireDate,LastDate;
	UBYTE CallsToday,Emulation,Expert,Rows;
	UWORD LastTime,Immortal;
	BYTE KeySequence[4];
	UBYTE Access;
	BYTE Class[19];
	BYTE Gang[20];
	BYTE Status[5],Blessed[5],Cursed[5],Coward;
	BYTE KeyHints[4][3];
	DOUBLE Experience;
	UBYTE Sex,Gender,Level,ExpLevel;
	UWORD HP,SP,Hull;
	UBYTE STR,INT,DEX,CHA;
	UBYTE Cannon,Ram;
	DOUBLE Gold,Bank,Loan;
	BYTE Weapon[34];
	BYTE Armor[34];
	BYTE WCmod,ACmod;
	UBYTE RealEstate,Security;
	UWORD Spell,Poison;
	struct charstats Current;
	struct playerstats History;
	DOUBLE Bounty;
	BYTE BountyWho[5];
	UBYTE Wins;
	DOUBLE Rating;
	UBYTE Novice;
	UBYTE XSpell;
	BYTE RemoteClient[24];
	UBYTE MyMaxSTR, MyMaxINT, MyMaxDEX, MyMaxCHA;
	UBYTE MyMelee, MyBackstab, MyPoison, MyMagic, MySteal;
};

struct userid {
	BYTE ID[5];
	UWORD record;
};

struct userlevel {
	BYTE Current[3];
	BYTE ID[5];
};

struct username {
	BYTE Handle[23];
	BYTE ID[5];
};

struct gang {
	BYTE Name[21];
	BYTE MemberID[4][5];
	UWORD Win,Loss;
	UBYTE BannerType,BannerFore,BannerBack,BannerTrim;
};

struct armor_item {
	BYTE Name[24];
	DOUBLE Value;
	UBYTE Class;
	BYTE EgoClass[10];
	BYTE EgoMelee;
	BYTE EgoMagic;
	BYTE EgoBearer[10];
	BYTE EgoCurse[10];
};
struct armor {
	BYTE Origin[9];
	UBYTE NumArmor;
	struct armor_item **Item;
};

struct magic {
	BYTE Name[20];
	UWORD Power[2];
	DOUBLE Cost[2];
};

struct poison {
	BYTE Name[20];
	UWORD Power[2];
	DOUBLE Cost[2];
};

struct realestate {
	BYTE Name[20];
	DOUBLE Value;
};

struct security {
	BYTE Name[20];
	DOUBLE Value;
};

struct weapon_item {
	BYTE Name[24];
	DOUBLE Value;
	UBYTE Class;
	BYTE *Action;
	BYTE EgoClass[10];
	BYTE EgoMelee;
	BYTE EgoMagic;
	BYTE EgoBearer[10];
	BYTE EgoCurse[10];
};
struct weapon {
	BYTE Origin[9];
	UBYTE NumWeapon;
	struct weapon_item **Item;
};

struct access {
	BYTE Name[16];
	UBYTE Calls,Minutes,Promote;
	UBYTE Verify,Bulletin,RolePlay,Weekend,Sysop;
};

struct arena {
	BYTE Name[20];
	BYTE Class[19];
	UBYTE Level;
	BYTE Weapon[34];
	BYTE Armor[34];
	DOUBLE Gold;
	UWORD Spell;
};

struct dungeon {
	BYTE Name[20];
	BYTE Class[19];
	UWORD Spell;
	UBYTE XSpell;
};

struct naval {
	BYTE Name[20];
	UBYTE INT;
	UBYTE Ram;
	UWORD Hull,Shot,Power;
	DOUBLE Value;
};

struct HiScore {
	DOUBLE HP,Hull,Kills;
	DOUBLE Calls,Killed,Retreats,KO,Joust,Stole,Gamble,Fish;
	DOUBLE Rating;
	BYTE Who[12][23];
};

struct ServerTable {
	UWORD NumAccess;
	struct access **access;
	UWORD NumArena;
	struct arena **arena;
	UWORD NumArmor;
	struct armor **armor;
	UWORD NumClass;
	struct class **class;
	UWORD NumDungeon;
	struct dungeon **dungeon;
	UWORD NumMagic;
	struct magic **magic;
	UWORD NumNaval;
	struct naval **naval;
	UWORD NumPoison;
	struct poison **poison;
	UWORD NumRealEstate;
	struct realestate **realestate;
	UWORD NumSecurity;
	struct security **security;
	UWORD NumWeapon;
	struct weapon **weapon;
	struct HiScore *HiScore;
	struct user *irsrec;
	struct user *newrec;
	struct user *sysrec;
};

struct RPGserver {
	BYTE client[5];
	BYTE command;
	UBYTE *parameter;
	UBYTE *result;
};
